---
title: "Admixture & Association Mapping Report"
author: Patrick Monnahan
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document:
    includes:
      in_header: header.tex
params:
  genesis_rslt: NA
  input_directory: NA
  blink_rslt: NA
  admixMap_rslt: NA
  ancestry_comps: NA
  global_ancestry: NA
  gwas_threshold: NA
  fam_file: NA
  gen_since_admix: NA
  sHWE_markers: NA
  stats_file: NA
  pruned_bim: NA
  ibd_ctrls: NA
  sHWE_entropy: NA
  eigenval: NA
  phenodat: NA
  kinplot: NA
  rulegraph_file: NA
  config_file: NA
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir=normalizePath('../'))
knitr::opts_chunk$set(echo = FALSE, fig.height = 6, fig.width = 6, fig.align = 'center', fig.pos = 'H')
# A Prefix nulling hook.

# Make sure to keep the default for normal processing.
default_output_hook <- knitr::knit_hooks$get("output")

# Output hooks handle normal R console output.
knitr::knit_hooks$set( output = function(x, options) {

  comment <- knitr::opts_current$get("comment")
  if( is.na(comment) ) comment <- ""
  can_null <- grepl( paste0( comment, "\\s*\\[\\d?\\]" ),
                     x, perl = TRUE)
  do_null <- isTRUE( knitr::opts_current$get("null_prefix") )
  if( can_null && do_null ) {
    # By default R print output aligns at the right brace.
    align_index <- regexpr( "\\]", x )[1] - 1
    # Two cases: start or newline
    re <- paste0( "^.{", align_index, "}\\]")
    rep <- comment
    x <- gsub( re, rep,  x )
    re <- paste0( "\\\n.{", align_index, "}\\]")
    rep <- paste0( "\n", comment )
    x <- gsub( re, rep,  x )
  }

  default_output_hook( x, options )

})

knitr::opts_template$set("kill_prefix"=list(comment=NA, null_prefix=TRUE))
knitr::opts_chunk$set(opts.label="kill_prefix")
```

```{r dependencies, include=FALSE}
library("ggplot2")
library("reshape2")
library("yaml")
library("dplyr")
library("tibble")
library("tidyr")
library("magrittr")
library("wrapr")
library("stringr")
library("foreach")
library("forcats")
library("ggridges")
library("GWASTools")
library("STEAM")
```

```{r, include=FALSE}
#Load functions
gg_qqplot <- function(ps, title, ci = 0.95) {
  n  <- length(ps)
  df <- data.frame(
    observed = -log10(sort(ps)),
    expected = -log10(ppoints(n)),
    clower   = -log10(qbeta(p = (1 - ci) / 2, shape1 = 1:n, shape2 = n:1)),
    cupper   = -log10(qbeta(p = (1 + ci) / 2, shape1 = 1:n, shape2 = n:1))
  )
  log10Pe <- expression(paste("Expected -log"[10], plain(P)))
  log10Po <- expression(paste("Observed -log"[10], plain(P)))
  ggplot(df) +
    geom_ribbon(
      mapping = aes(x = expected, ymin = clower, ymax = cupper),
      alpha = 0.1
    ) +
    geom_point(aes(expected, observed), shape = 1, size = 3) +
    geom_abline(intercept = 0, slope = 1, alpha = 0.5) +
    # geom_line(aes(expected, cupper), linetype = 2, size = 0.5) +
    # geom_line(aes(expected, clower), linetype = 2, size = 0.5) +
    xlab(log10Pe) +
    ylab(log10Po) + ggtitle(title)
}

```


```{r read_params, include=FALSE}

if(any(unlist(lapply(params, function(x) x=="NA")))) stop("Missing input params")

fam = read.table(params$fam_file, comment.char = "")
cases = fam %>% filter(V6==2) %>% select(V2) %>% mutate(V1=V2) %>% select(V1)

rulegraph_file <- params$rulegraph_file
yaml <- read_yaml(params$config_file)

gen = read.table(params$genesis_rslt, header = T)
blink = read.table(params$blink_rslt, header = T)
if (params$admixMap_rslt != "-9"){
  adm = read.table(params$admixMap_rslt, header = T)
}
pbim = read.table(params$pruned_bim)
anc = read.table(params$global_ancestry, header=T)
Gens = as.numeric(params$gen_since_admix)

ancpre = strsplit(params$ancestry_comps, ",")[[1]]

gwas_thresh = as.numeric(params$gwas_threshold)

ent = read.table(params$sHWE_entropy)
shwe = read.table(params$sHWE_markers)
ibd_ctrl = read.table(params$ibd_ctrls)
eigval = read.table(params$eigenval, skip=1)
eigval %<>% filter(!is.na(V1)) %>% rowid_to_column %>% mutate(PC = rowid, VarianceProportion = V1) %>% select(PC, VarianceProportion)
pdat = read.table(params$phenodat, head=T)

outPre = str_split(params$blink_rslt, "[.]")[[1]][1]

```

# Summary

This report summarizes relevant information of the input, output, and intermediate steps of a pipeline (\href{https://github.com/pmonnahan/admixMap}{{\color{blue}{\underline{Github repo}}}}) which performs both admixture and association mapping.  Prior to mapping, input data is parsed and QC'ed to identify (and optionally filter) related samples, calculate principal components, and perform genomic control matching.  

## Genome Wide Association Mapping (GWAS)
GWAS is performed via two softwares: \href{http://zzlab.net/blink/}{{\color{blue}{\underline{BLINK}}}} and \href{https://github.com/UW-GAC/GENESIS}{{\color{blue}{\underline{GENESIS}}}}.  BLINK implements a novel multi-locus methodology that is fast, capable of handling large datasets, and claims greater statistical power.  However, the program is incapable of generating coefficient estimates necessary for calculating odds ratios, polygenic risk scores, etc.  GENESIS scales poorly with sample size, but is well-documented, written in R (and thus more transparent), and provides much more detailed output.

The following plink file were used as the initial input for the GWAS portion of this pipeline:

```{r input-datasets}
knitr::opts_chunk$set(opts.label="kill_prefix")
yaml$query
```

## Admixture Mapping
For the admixture mapping, this pipeline takes, as input, the output of the \href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3738819/}{{\color{blue}{\underline{RFMix}}}} ancestry inference program. The model used for admixture mapping is taken from (\href{https://www.cell.com/ajhg/pdfExtended/S0002-9297(19)30008-4}{{\color{blue}{\underline{Grinde et al 2019}}}}).  In short, for every genomic window a general linear model (generalized, in the case of case/control phenotypes) is fit with the local ancestry state as the predictor variable with global ancestry (or admixture proportions) as a covariate.  For every ancestry of interest, a separate model is fit, which corresponded to the following ancestries, in this case:

```{r ancestries}
yaml$ancestry_predictors
```

The following additional covariates were also included:

```{r covars}
yaml$other_predictors
```

The RFMix results used for admixture mapping were taken from:
```{r rfmix-results}
yaml$admixMapping$rfmix
```

And, the raw output of the \href{https://github.com/pmonnahan/admixMap/blob/master/scripts/admixMap2.R}{{\color{blue}{\underline{admixture mapping script}}}} can be found here:
```{r admixmap-results}
paste(getwd(), params$admixMap_rslt, sep="/")
```

The \href{https://www.cell.com/ajhg/pdfExtended/S0002-9297(19)30008-4}{{\color{blue}{\underline{Grinde et al 2019}}}} paper also provides a framework and software (called \href{https://github.com/kegrinde/STEAM}{{\color{blue}{\underline{STEAM}}}}) for determining an appropriate significance threshold for multiple test correction. It is based on the genetic positions of the tested windows, the average ancestry of each individual, and the number of generations since admixture began.  Using the window positions in the output listed above, the global ancestries found here

```{r global-ancestry}
paste(getwd(), params$global_ancestry, sep="/")
```

and assuming admixture began 

```{r gen_since_admix}
params$gen_since_admix
```
generations ago, the software determined the following 5% significance threshold 

### Significance Threshold
```{r admixMap-sig-thresh, warning=FALSE, message=FALSE}

if (params$admixMap_rslt != "-9"){
  dat = adm %>% mutate(cM = (sgpos + egpos) / 2, chr = chm) %>% distinct(cM, chr) 
  anc2 = anc %>% select(-sample)

  thresh = get_thresh_simstat(g = Gens, map = dat, props = anc2, nreps = 1000)
  print(thresh$threshold)
  adm_thresh = thresh$threshold
  
  adm %>% filter(ANC.p.value < thresh$threshold) %>% write.table(paste0(outPre,".admixmap.sig.txt"), row.names = F, quote=F)
}
```


# Quality Control
The first step in the pipeline is to subset the input PLINK files (and admixture mapping input) listed above with an optionally provided list from the user.  

Sample file:
```{r sample-file}
yaml$samples
```

The resulting dataset is composed of the following numbers of cases and controls:
```{r case-numbers-sub}
fam %>% mutate(phenotype = V6) %>% group_by(phenotype) %>% summarize(n=n()) %>% knitr::kable()
```


Then, the PLINK files are run through a series of of QC steps in order to prepare the data for calculation of PCs.  The first step is to isolate the control subjects and prune markers showing high amounts of linkage disequilibrium (based on _r^2^_) in the controls.  

The parameters used for LD pruning in PLINK were:
```{r LD-pruning-params}
yaml$LD_pruning
```

PLINK is then used to calculate Identity by Descent (IBD) for each pair of control samples, and individuals are removed such that no two individuals remain with a 'PI_HAT' value above:
```{r IBD-thresh}
yaml$IBD$pi_hat
```

, which resulted in a dataset with the following number of unrelated controls:
```{r IBD-ctrl-num}
nrow(ibd_ctrl)
```


## Structured Hardy-Weinberg Filter
The sHWE (\href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6827367/}{{\color{blue}{\underline{structured Hardy-Weinberg equilibrium}}}}) method facilitates testing of HWE when a dataset contains multiple, genetically-structured populations.  However, there are two substantial limitations to this approach.  First, it requires the user to specify the number of populations in the dataset.  Since this is generally an unknown parameter, the authors suggest calculating p-values under a range of population numbers and then using the maximum entropy of the p-value distribution as criteria for finding the optimal population number.  The entropy is calculated over all bins of the p-value distribution excluding the left-most bin which contains the true SNPs violating HWE.  The remainder of this distribution should be uniform if the population parameter is accurate.  Entropy increases as the distribution becomes more uniform, so max entropy is used as the criteria.

Entropy values were calculated from p-value distributions generated assuming the following range of population number:

```{r sHWE-entropy, fig.height = 4, fig.width = 4, fig.cap = "Entropy of p-value distributions (excluding the first bin) for varying population number", warning=FALSE, message=FALSE}
ent %>% ggplot(aes(x=V1,y=V2)) + geom_point() + geom_line() + xlab("Number of Populations") + ylab("Entropy")
```

The population value with the maximum entropy is:
```{r sHWE-opt-pop}
ent %>% filter(V2==max(ent$V2)) %>% select(V1)
```


The other main limitation to the program is that it does not allow any missing data at a marker.  Since markers with no missing data could represent a biased subset of the genotype data, the pipeline performs a large series of downsampling iterations (with user-specified parameters) with the hope that, by random chance, a subset of the markers with missing data in the full dataset will have no missing data in the downsampled dataset(s).  

In this case, the pipeline generated:
```{r sHWE-downsampling-rounds}
yaml$sHWE$downsample_rounds
```

downsampled datasets, each containing:

```{r sHWE-downsampling-inds}
yaml$sHWE$downsample_number
```
individuals.  

Using lower numbers of individuals in each downsampled subset will increase the likelihood that any one marker will have complete genotype information, but going too low will not provide a reliable sample of the larger control population.  It is recommended that the user decide on an acceptable downsample size and increase the number of downsampling rounds until the desired number or proportion of markers are tested.

The minimum number of tested markers requested was:
```{r sHWE-minimum-tests}
yaml$sHWE$test_threshold
```

, and the realized number of tested markers was:
```{r sHWE-realized-tests}
nrow(shwe)
```


## Population structure and Kinship

The unrelated samples used in PC calculation (via PC-AiR) can be found here:
```{r PCair-unrelateds}
paste(getwd(), str_replace(params$eigenval, "varprop", "unrels"), sep = "/")
```

, and the related samples based on a kinship value greater than
```{r PCair-kinship}
yaml$kinship_threshold
```

can be found here:
```{r PCair-relateds}
paste(getwd(), str_replace(params$eigenval, "varprop", "rels"), sep = '/')
```

PC-AiR Variance Proportions:
```{r plink-eigenvalues, fig.height = 4, fig.width = 4, fig.cap = "Proportion of variance explained by each PC from PC-AiR", warning=FALSE, message=FALSE}
eigval %>% ggplot(aes(x = PC, y = VarianceProportion)) + geom_bar(stat="identity") + ylab("Variance Proportion") + xlab("PC")
```

PC plots:
```{r plink-pcplots, fig.height = 4, fig.width = 4, fig.cap = "Pairwise PC plots (via PC-AiR) for the number of PCs used in GWAS", warning=FALSE, message=FALSE}
for (i in 1:as.numeric(yaml$gwas$pc_num)){
  for (j in (i+1):as.numeric(yaml$gwas$pc_num)){
    xvar = paste0(".$PC",i)
    yvar = paste0(".$PC",j)
    pdat %>% mutate(Pheno = as.character(pheno)) %.>% print(ggplot(data = .,aes_string(x=xvar,y=yvar,color='sex',shape='Pheno')) + geom_point(alpha = 0.5))
  }
}
```

Kinship was recalculated via PCRelate, resulting in the following distribution of pairwise kinship values (y-axis):
```{r, fig.height = 4, fig.width = 4, fig.cap = "Pairwise kinship plotted vs the probability that the pair of individuals have 0 alleles IBD at a random site on the genome (k0)", warning=FALSE, message=FALSE}
knitr::include_graphics(normalizePath(params$kinplot))
```

There should be a negative relationship between kinship and k0.  

## Control Matching
Control matching is optionally performed for BLINK and admixture mapping.  Logistic models implemented in GENESIS are able to handle highly imbalanced case/control ratios by way of a Saddle-Point Approximation for p-values, so control matching is not necessary.  

Was control matching performed?
```{r, ctrl_match, warning=FALSE, message=FALSE}
yaml$match_controls
```

If 'true', then PLINK's \href{http://www.cog-genomics.org/plink/1.9/strat}{{\color{blue}{\underline{control matching program}}}} was implemented while requested the following number of controls to be retained per case sample (via -mcc flag):
```{r, ctrl_num, warning=FALSE, message=FALSE}
yaml$control_number
```

The retained samples can be found here:
```{r, matched_samples, warning=FALSE, message=FALSE}
paste(getwd(), str_replace(params$fam_file, "_sub", "_CtlMat"), sep="/")
```


# Results

## GWAS

```{r, warning=FALSE, message=FALSE}
#Output significant SNPs
blink %>% filter(p_value <= gwas_thresh) %>% write.table(paste0(outPre,".blink.sig.txt"), row.names = F, quote=F)
gen %>% filter(SPA.pval <= gwas_thresh) %>% write.table(paste0(outPre,".genesis.sig.txt"), row.names = F, quote=F)
```

### GENESIS
Implementation of the GENESIS software generally followed the provided \href{https://www.bioconductor.org/packages/release/bioc/vignettes/GENESIS/inst/doc/assoc_test.html}{{\color{blue}{\underline{vignette}}}}.  In brief, the LD-pruned SNPs explained above are first used to calculate a first-pass kinship matrix via \href{http://people.virginia.edu/~wc9c/KING/manual.html}{{\color{blue}{\underline{KING}}}}, which is designed to be robust in the face of population structure.  Then, these kinship estimates along with the LD-pruned SNPs are used to calculate ancestry-informative PCs via \href{https://rdrr.io/bioc/GENESIS/man/pcair.html}{{\color{blue}{\underline{PC-AiR}}}}. The basic idea behind PC-AiR is to calculate PCs using only unrelated individuals and then use the resulting eigenvectors/values to project PC values for the related individuals that had been excluded. These PCs are formatted and combined with the phenotype and covariate data, all of which is available here:

```{r genesis-pdat}
paste(getwd(), str_replace(params$genesis_rslt,"txt","pdat"), sep="/")
```

Finally, a genetic-relatedness matrix is calculated via \href{https://www.rdocumentation.org/packages/GENESIS/versions/2.2.2/topics/pcrelate}{{\color{blue}{\underline{PC-Relate}}}}, which 'accounts for population structure (ancestry) among sample individuals through the use of ancestry representative principal components (PCs)'. 

For the GWAS, the following number of PCs from PC-AiR were used to control for population structure:
```{r pc-num2}
yaml$gwas$pc_num
```
and the genetic-relatedness matrix was used for the error-covariance matrix.

The raw output of the GENESIS GWAS can be found here:
```{r genesis-output}
paste(getwd(), params$genesis_rslt, sep = "/")
```
and the meaning of each column name can be found \href{https://rdrr.io/bioc/GENESIS/man/assocTestSingle.html}{{\color{blue}{\underline{here}}}}

SNPs with p-values below the following thresholds were annotated via \href{http://snpeff.sourceforge.net/SnpEff_manual.html}{{\color{blue}{\underline{SnpEff}}}}:
```{r gen-sig-thresh}
yaml$gwas$sig_threshold
```

and output to:
```{r genesis-annt-snps}
paste(getwd(), str_replace(params$genesis_rslt, "txt", "annt.txt"), sep = "/")
```

```{r, fig.height = 4, fig.width = 4, fig.cap = "Quantile-Quantile (QQ) plot for a random sample of (500k) p-values resulting from the GENESIS GWAS", warning=FALSE, message=FALSE}
knitr::opts_chunk$set(dev = 'png')
gen %>% sample_n(500000) %>% filter(!is.na(SPA.pval)) %.>% gg_qqplot(.$SPA.pval, "") 
```

and a corresponding Genomic Inflation Factor (GIF) of:
```{r genesis-inflation-factor}
gen %>% filter(variant.id %in% pbim$V2) %>% mutate(chisq = Score.Stat ^ 2) %>% select(chisq) %.>% median(.$chisq, na.rm=T) / 0.456
```
A GIF near of 1 indicates absence of p-value inflation due to population structure, and a value < 1.1 is generally considered acceptable.  This GIF is based on the test statistics of the LD-pruned SNPs and \href{http://rstudio-pubs-static.s3.amazonaws.com/9743_8a5f7ba3aa724d4b8270c621fdf6d06e.html}{{\color{blue}{\underline{calculated}}}} by dividing the median of the squared test statistics by 0.456.  

```{r, fig.height = 5, fig.width = 7, fig.cap = "Manhattan plot of log(10) of p-values produced from the GENESIS program", warning=FALSE, message=FALSE}
knitr::opts_chunk$set(dev = 'png')
manhattanPlot(gen$SPA.pval, gen$chr)
```

```{r, include=F}
png("figures/genesis.mnht.png")
manhattanPlot(gen$SPA.pval, gen$chr)
dev.off()
```

### BLINK
BLINK (Bayesian-information and Linkage-disequilibrium Iteratively Nested Keyway) takes a substantially different approach to GWAS than GENESIS.  As opposed to GENESIS, which tests a single marker at a time, the BLINK procedure is inherently multi-locus, using Bayesian information criteria to determine which loci to include in the model.  Via this model selection process, it eliminates the need to estimate random effects related to kinship. Population structure is corrected for via PCs calculated by PLINK, using the following number of PCs:

```{r pc-num}
yaml$gwas$pc_num
```

PLINK was used to calculate PCs because using the PCs from PC-AiR described above resulted in inflation of p-values according to QQ plots.  

The raw output for BLINK were output to:
```{r blink-output}
paste(getwd(), params$blink_rslt, sep="/")
```

SNPs with p-values below the following thresholds were annotated via \href{http://snpeff.sourceforge.net/SnpEff_manual.html}{{\color{blue}{\underline{SnpEff}}}}:
```{r blink-sig-thresh}
yaml$gwas$sig_threshold
```

and output to:
```{r blink-annt-snps}
paste(getwd(), str_replace(params$blink_rslt, "txt","annt.txt"), sep="/")
```

```{r, fig.height = 4, fig.width = 4, fig.cap = "Quantile-Quantile (QQ) plot for a random sample of (500k) p-values resulting from the BLINK GWAS", warning=FALSE, message=FALSE}
knitr::opts_chunk$set(dev = 'png')
blink %>% sample_n(500000) %>% filter(!is.na(p_value)) %.>% gg_qqplot(.$p_value, "") 
```
Note: Since BLINK does not output any type of test statistic (which serves as the basis of the Inflation Factor), it is not possible to calculate the GIF.

```{r, fig.height = 5, fig.width = 7, fig.cap = "Manhattan plot of log(10) of p-values produced from the BLINK program", warning=FALSE, message=FALSE}
knitr::opts_chunk$set(dev = 'png')
manhattanPlot(blink$p_value, blink$chr)
```

```{r, include=F}
png("figures/blink.mnht.png")
manhattanPlot(blink$p_value, blink$chr)
dev.off()
```

## Admixture Mapping

```{r, fig.height = 4, fig.width = 4, fig.cap = "Quantile-Quantile (QQ) plot for a random sample of (500k) p-values for an effect of local ancestry", warning=FALSE, message=FALSE}
for (i in 1:length(ancpre)){
  adm %>% filter(anc==ancpre[i] & !is.na(ANC.p.value)) %.>% print(gg_qqplot(.$ANC.p.value, ancpre[i]))
}
```

```{r, fig.height = 4, fig.width = 7, fig.cap = "Manhattan plot of log(10) of p-values produced from the BLINK program", warning=FALSE, message=FALSE}
knitr::opts_chunk$set(dev = 'png')
for (i in 1:length(ancpre)){
  print(ancpre[i])
  adm %>% filter(anc==ancpre[i] & !is.na(ANC.p.value)) %>% mutate(Chr = as.numeric(chm)) %>% arrange(Chr) %.>% manhattanPlot(.$ANC.p.value, .$chm, signif = adm_thresh)
}
```


```{r, include=F}
for (i in 1:length(ancpre)){
  png(paste0("figures/admixMap.",ancpre[i], ".mnht.png"))
  adm %>% filter(anc==ancpre[i] & !is.na(ANC.p.value)) %.>% manhattanPlot(.$ANC.p.value, .$chm, signif = adm_thresh)
  dev.off()
}
```


```{r}
# for (i = 1:22){
#   png()
# blink.slim %>% filter(chr==10 & datset!='all')%>% ggplot(aes(y=-log10(p_value), x = pos/1000000, color = datset)) + geom_point(alpha=0.3) + 
# geom_hline(yintercept = -log10(0.05), linetype="dashed") + geom_hline(yintercept = -log10(0.00000001), linetype="dotted") + geom_vline(xintercept = 63.723577, linetype="dashed", color = "red", alpha = 0.5) + geom_vline(xintercept = 8.104208, linetype="dashed", color = "red", alpha = 0.5) + geom_vline(xintercept = 22.856946, linetype="dashed", color = "red", alpha = 0.5) + geom_vline(xintercept = 12.6292655, linetype="dashed", color = "red", alpha = 0.5) + geom_text(x = 69.823577, y = 20, label="ARID5b") + geom_text(x = 2.104208, y = 20, label="GATA3") + geom_text(x = 29.856946, y = 18, label="PIP4K2A") + geom_text(x = 17.6292655, y = 22, label="LHPP") + ggtitle("Chromosome 10") + xlab("Position (Mb)")
# }

```





```{r rfmix-chrom.plots, include=FALSE}
# Consider adding centromere locations
if (adm != "-9"){
  files = list.files(params$input_directory, pattern = ".dat")
  for (f in 1:length(files)){
    in_name = paste(params$input_directory, files[f], sep = "/")
    chrom = str_split(files[f], '[.]')[[1]][[1]]
    # print(in_name)
    dat = read.table(in_name, header = T)
    png(paste0("figures/chrom_plots/",chrom,".png"))
    plt = dat %>% mutate(pos = (spos + epos) / 2) %>% select(c("pos","pheno", ancpre)) %>% pivot_longer(-c(pos,pheno), names_to="Ancestry") %>% group_by(pos,pheno,Ancestry) %>% summarize(Tot = sum(value)) %>% mutate(prop = Tot / sum(Tot)) %>% ggplot(aes(x = pos / 1000000, y = prop, linetype=as.character(pheno), color = Ancestry)) + geom_line() + xlab("Position (Mb)") + ylab("Proportion of Haplotypes") + scale_linetype_discrete(name = "Phenotype", breaks = c(0,1), labels = c("Control", "Case"))
    print(plt)
    dev.off()
  }
}

```



\newpage

## Rule Graph

Below is a directed acyclic graph depicting the steps involved in this post-imputation QC pipeline.  When possible, computation within each node was parallelized by dataset, chromosome, etc.  The full DAG visualizing the parallel computing can be generated via:

    snakemake --dag | dot -Tpng > jobgraph.png
    
from within the directory that the post-imputation QC was carried out.  These are typically too large to fit easily in a pdf, and so were not included in this report.

```{r, fig.cap = "A rule graph showing the different steps of the bioinformatic analysis that is included in the Snakemake workflow.", out.height = "11cm"}
knitr::include_graphics(normalizePath(rulegraph_file))
```

\newpage

## Reproducibility

The code for reproducing this analysis is available \href{https://github.com/pmonnahan/admixMap}{{\color{blue}{\underline{here}}}}. The repo contains:

* A Snakemake workflow for running all steps.
* A collection of scripts to acheive individual steps
* A Singularity definitions file that can be used to generate the Singularity image used to run all steps.
** This image file is also directly available upon request

The pipeline was carried out using the following singularity image:
```{r singularity-image2}
yaml$singularity$image
```


The code for reproducing this report is available \href{https://github.com/pmonnahan/admixMap/blob/master/scripts/admixMap_report.Rmd}{{\color{blue}{\underline{here}}}}.

The input files for the figures produced herein are from:

```{r input-files}
params
```
Also, see the config.yml in the workflow directory for full list of parameter inputs and settings.

The results in this supplementary were generated in the following R environment:

\footnotesize
```{r session_info}
sessionInfo()
```
\normalsize



